libs/application/ui-shell/src/reducer/reducerUtils.ts

```diff
@@ -32,50 +33,129 @@ import {
+
+    if (!isNavigable || doesNotRequireAnswer) {
+      let skipLength = 1
+
+      if (
+        screen.type === FormItemTypes.MULTI_FIELD ||
+        screen.type === FormItemTypes.REPEATER
+      ) {
+        skipLength = screen.children.length + 1
+      }
+
+      currentScreen = index + skipLength
+      index = index + skipLength
+
+      continue
+    }
 
-  screens.forEach((screen, index) => {
     if (screen.type === FormItemTypes.MULTI_FIELD) {
       let numberOfAnsweredQuestionsInScreen = 0
 
-      screen.children.forEach((field) => {
-        if (getValueViaPath(answers, field.id) !== undefined) {
+      for (const subScreen of screen.children) {
+        const isSubScreenNavigable = get(subScreen, 'isNavigable') === true
+        const subScreenDoesNotRequireAnswer =
+          get(subScreen, 'doesNotRequireAnswer') === true
+        const subScreenHasBeenAnswered =
+          getValueViaPath(answers, subScreen.id) !== undefined
+
+        if (!isSubScreenNavigable || subScreenDoesNotRequireAnswer) {
+          numberOfAnsweredQuestionsInScreen++
+        } else if (subScreenHasBeenAnswered) {
           numberOfAnsweredQuestionsInScreen++
+        } else {
+          break
         }
-      })
+      }
 
-      if (numberOfAnsweredQuestionsInScreen === screen.children.length) {
-        currentScreen = index + 1
-      } else if (numberOfAnsweredQuestionsInScreen > 0) {
+      const answeredAllSubScreens =
+        numberOfAnsweredQuestionsInScreen === screen.children.length
+
+      if (answeredAllSubScreens) {
+        lastAnswerIndex = index + 1
+        currentScreen = index + screen.children.length
+      } else {
+        // Did not answer all the questions
         currentScreen = index
+        lastAnswerIndex = currentScreen
+        missingAnswer = true
+
+        if (stopOnFirstMissingAnswer) {
+          break
+        }
       }
     } else if (screen.type === FormItemTypes.REPEATER) {
-      if (getValueViaPath(answers, screen.id) !== undefined) {
+      const repeaterAnswer = getValueViaPath(answers, screen.id)
+      const repeaterHasAnswer = repeaterAnswer !== undefined
+
+      if (
+        repeaterHasAnswer &&
+        isArray(repeaterAnswer) &&
+        repeaterAnswer.length > 0
+      ) {
+        lastAnswerIndex = index
         currentScreen = index
+      } else {
+        missingAnswer = true
+
+        if (stopOnFirstMissingAnswer) {
+          break
+        }
       }
     } else if (screen.id) {
-      if (
-        getValueViaPath(answers, screen.id) === undefined &&
-        index > currentAnswerIndex
-      ) {
-        missingAnswerBeforeCurrentIndex = true
-        currentAnswerIndex = index
+      const screenHasBeenAnswered =
+        getValueViaPath(answers, screen.id) !== undefined
+
+      if (!screenHasBeenAnswered && stopOnFirstMissingAnswer) {
+        currentScreen = index
+        break
       }
 
-      if (
-        !missingAnswerBeforeCurrentIndex &&
```

Comment:
Doesn't this check (and `missingAnswer` in the new code) have the same behaviour as `stopOnFirstMissingAnswer`? Why does it keep going?